cmd/restic/cmd_prune.go

```diff
@@ -82,226 +123,400 @@ func runPrune(gopts GlobalOptions) error {
-		bar.Start()
-		obsoletePacks, err = repository.Repack(ctx, repo, rewritePacks, usedBlobs, bar)
-		if err != nil {
-			return err
+		//sort repackCandidates such that packs with highest ratio unused/used space are picked first
+		sort.Slice(repackCandidates, func(i, j int) bool {
+			return repackCandidates[i].unusedSize*repackCandidates[j].usedSize >
+				repackCandidates[j].unusedSize*repackCandidates[i].usedSize
+		})
+
+		for i, p := range repackCandidates {
+			if stats.size.unused < maxUnusedSizeAfter+stats.size.remove+stats.size.repackrm {
+				stats.packs.keepPartlyUsed += uint(len(repackCandidates) - i)
+				break
+			}
+			repack(p.ID, p.packInfo)
 		}
-		bar.Done()
+	} else {
+		stats.packs.keepPartlyUsed += uint(len(repackCandidates))
 	}
 
-	removePacks.Merge(obsoletePacks)
+	Verbosef("\nused:        %10d blobs / %s\n", stats.blobs.used, formatBytes(stats.size.used))
+	Verbosef("duplicates:  %10d blobs / %s\n", stats.blobs.duplicate, formatBytes(stats.size.duplicate))
+	Verbosef("unused:      %10d blobs / %s\n", stats.blobs.unused, formatBytes(stats.size.unused))
+	Verbosef("unreferenced:                   %s\n", formatBytes(stats.size.unref))
+	totalSize := stats.size.used + stats.size.duplicate + stats.size.unused + stats.size.unref
+	Verbosef("total:       %10d blobs / %s\n", stats.blobs.used+stats.blobs.unused+stats.blobs.duplicate,
+		formatBytes(totalSize))
+	Verbosef("unused size: %s of total size\n\n", formatPercent(stats.size.unused, totalSize))
+
+	Verbosef("to repack:   %10d blobs / %s\n", stats.blobs.repack, formatBytes(stats.size.repack))
+	Verbosef("  -> prunes: %10d blobs / %s\n", stats.blobs.repackrm, formatBytes(stats.size.repackrm))
+	Verbosef("to delete:   %10d blobs / %s\n", stats.blobs.remove, formatBytes(stats.size.remove))
+	Verbosef("delete unreferenced:            %s\n", formatBytes(stats.size.unref))
+	totalPruneSize := stats.size.remove + stats.size.repackrm + stats.size.unref
+	Verbosef("total prune: %10d blobs / %s\n", stats.blobs.remove+stats.blobs.repackrm, formatBytes(totalPruneSize))
+	Verbosef("unused size after prune: %s of total size\n\n",
+		formatPercent(stats.size.unused-stats.size.remove-stats.size.repackrm, totalSize-totalPruneSize))
+
+	Verbosef("total data files: %d / keep used: %d, keep partly used: %d, repack: %d, delete: %d, delete unreferenced: %d\n\n",
+		int(stats.packs.keepUsed+stats.packs.keepPartlyUsed)+len(repackPacks)+len(removePacks)+len(removePacksFirst),
+		stats.packs.keepUsed, stats.packs.keepPartlyUsed, len(repackPacks), len(removePacks), len(removePacksFirst))
+
+	// unreferenced packs can be safely deleted first
+	if len(removePacksFirst) != 0 {
+		Verbosef("deleting unreferenced data files...\n")
+		DeleteFiles(gopts, opts.DryRun, true, repo, removePacksFirst, restic.DataFile)
+	}
 
-	if err = rebuildIndex(ctx, repo, removePacks); err != nil {
-		return err
+	if len(repackPacks) != 0 {
+		if !opts.DryRun {
```

Comment:
I think the dry run case should be handled first, so that the regular case can be unindented.

Reply:
Can you please specify what you mean? Of course I can change the order of the `if` and handle the dry run case fist, but this will still mean the regular case is still indented in the `else` branch...

Reply:
I think what meant is, if you're already returning inside dry run blocks, you won't need an `else` case, so you can remove indentation for the rest of the code.

Reply:
Ok, the point is, there is no return inside this dry run block. In fact the code for deletion should also run in the dry run case (with submitting the flag).
I'll mark this as resolved. Please feel free to reopen if you have a good proposal.

Reply:
The dryrun case can end with a continue. Then the common case can be unindented.

Reply:
Sorry, I might be just too stupid to understand your point. As specified in the [language reference](https://golang.org/ref/spec#Continue_statements) a continue can only be used within a for loop. At this point in code, there is no for loop...

Reply:
I'm the one being stupid. I thought this was in a loop; I likely confused it with the other case, below. The only way to improve this stylistically is to put the dryrun case first because it's shorter and the condition need not be negated. It's not a big deal, though, if it doesn't skip a level of indentation.